_posts/2020-04-03-rendering-in-UE4.md

```diff
@@ -0,0 +1,278 @@
+
+Then, GPU starts to render object by object, firing a *draw call*. This consist of a group of polygons that shares the same properties. Between every drawcall, the render state could change (meaning that the render needs to set the different properties for the actual draw call). Behind the scene, Unreal sorts the drawcalls to optimize the number of state changes. The fewer changes the engine needs to do, the better are the performances. Usually, this optimization consist in grouping drawcalls by materials, and the engine does this automatically.
+
+Drawcalls have a huge impact on performance: every time the render is done executing the commands, it needs to receive the new commands for the next call from the render thread and this adds a significant overhead.
+
+With `stat RHI` you can check the number of drawcalls that are called from the renderer. Broadly (because it depends on a lot of different variables) a reasonable number is 2000/3000 for a desktop application. 5000 is still acceptable for good hardware; over 10000 it's probably a problem. On mobile, a good number is far lower, about a few hundred.
+
+Another very useful - and I would say *essential* - tool is [RenderDoc](https://renderdoc.org/): it's a free standalone graphics debugger that allows quick and easy single-frame capture of a single frame to make a detailed inspection of all the calls the application does. It works with every render API. Unreal has a plugin for it, see more [here](https://docs.unrealengine.com/en-US/Engine/Performance/RenderDoc/index.html).
+
+Here is a screenshot of a capture of a simple scene in Unreal:
+
+<a href="/images/graphics/FrameInRenderDoc.png"><img class="blog-image" src="/images/graphics/FrameInRenderDoc.png"> </a>
+
+On the left, you can see the *Event browser* that lists all the graphics API commands. In the *Scene* section there are the render passes, and if you expand a pass you'll find the list of draw calls for that step. On the right, there are the outputs of the selected pass: you can see the depth, and some of the GBuffer buffers filled in.
+As you can see on the left of the window, Unreal does a lot of passes for a frame. Even if you have 5 objects in the scene (the floor, the sky, and the three framed elements) the engine fires a lot of draw calls for all the other different passes.
+
+This is a very big topic, so for drawcalls I would finish here. I highly suggest to use *RenderDoc* to see how a frame is composed.
+
+The thing to pay attention is that in real-time rendering objects are rendered one by one, and for the Unreal Engine renderer an object is a primitive component, not an actor. This means that when you have a blueprint with 5 static mesh components in it with a single material each, that will results in 5 draw calls (at least, it's actually 5 drawcalls for every pass the object is rendered into). Plus, if you have a single static mesh component with two materials assigned, then you'll have 2 drawcalls, one per material.
```

Comment:
"The thing to pay attention is that in real-time rendering objects are rendered one by one, and for the Unreal Engine renderer an object is a primitive component, not an actor. "

This statement is not clear